rts/Lua/LuaSyncedRead.cpp

```diff
@@ -8515,8 +8520,139 @@ int LuaSyncedRead::GetUnitScriptNames(lua_State* L)
+
+			if (CCollisionHandler::DetectHit(u, u->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, u->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayFeatures
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @treturn {{len, Id},...}
+ */
+int LuaSyncedRead::TraceRayFeatures(lua_State* L) //returns the list of features that an raytrace has hit
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	QuadFieldQuery qfQuery;
+	quadField.GetQuadsOnRay(qfQuery, pos, dir, traceLength);
+	CollisionQuery cq;
+	int num = 0;
+	lua_newtable(L);
+	for (const int quadIdx : *qfQuery.quads) {
+		const CQuadField::Quad& quad = quadField.GetQuad(quadIdx);
+
+		for (CFeature* f : quad.features) {
+			if (!f->HasCollidableStateBit(CSolidObject::CSTATE_BIT_QUADMAPRAYS))
+				continue;
+
+			if (CCollisionHandler::DetectHit(f, f->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, f->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayGround
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @bool[opt=false] ignoreWater
+ * @treturn nil|len
+ * @treturn number     posX
+ * @treturn number     posY
+ * @treturn number     posZ
+ */
+int LuaSyncedRead::TraceRayGround(lua_State* L)
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	float3 normDir = dir.Normalize();
+	auto result2 = pos + normDir * traceLength;
+	const float groundLength = CGround::LineGroundCol(pos, result2);
+	const float waterRayLength = CGround::LinePlaneCol(pos, normDir, traceLength, 0.0f);
+	const int optArgIdx = 8 + lua_isnumber(L, 8);
+	const bool ignoreWater = luaL_optboolean(L, optArgIdx, false);
+	const float height = CGround::GetHeightReal(pos + normDir * traceLength);
+	if (-1.0f == groundLength)
+		return 0;
+	if (traceLength > groundLength && groundLength > 0.0f) {
+		traceLength = groundLength;
+	}
+	auto realvalue = CGround::LinePlaneCol(pos, normDir, traceLength, height);
+	if (-1.0f != realvalue && std::numeric_limits<float>::max() != realvalue)
+		traceLength == realvalue;
```

Comment:
why? what does this do? two calls to CGround::LineFooCol (one for ground, one for water) should be sufficient

Reply:
Was attempting to get it to follow Los. Will Remove that.